core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -178,6 +178,7 @@ public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, Sym
         private final TypeProvider types;
         private final ExpressionEquivalence expressionEquivalence;
         private final boolean dynamicFiltering;
+        private static final int MAX_LINEAR_EXPANSION_LENGTH = 1024;
```

Comment:
this seems pretty high. It means I could explode:
```
x := x + x
x := x + x
x := x + x
```
to 1000 terms pretty quickly. What is a use case where it helps?

Reply:
This constant can be negotiable, but at least we should be able to handle a dozen predicates with reasonable lengths, which is a common use case, e.g. ```where partition_date = DATE_PARSE('2021-07-01', '%Y-%m-%d') OR partition_date = DATE_PARSE('2021-07-012, '%Y-%m-%d')...```
Maybe we can put a limit on "count" (e.g. < 20) to prevent the case for your concern?

Reply:
If `where partition_date = DATE_PARSE('2021-07-01', '%Y-%m-%d') OR partition_date = DATE_PARSE('2021-07-012, '%Y-%m-%d')...`
is your use case, then I think we should simply translate that into IN instead of relaxing inlining conditions.
See https://github.com/trinodb/trino/issues/1118

Reply:
Thanks for the pointer. It looks like #2698 has no update since last November. More importantly, I think that's orthogonal to this issue. A series of equality conditions is just one use case. There are other cases like ```where partition LIKE 'ab%' OR partition LIKE 'cd%'...``` or inequality conditions like ```where partition BETWEEN DATE '2000-01-01' AND '2000-12-31' OR partition BETWEEN '2020-01-01' AND '2020-12-31'... ``` So I think it makes sense to relax this rule for all the common cases

Reply:
@weiatwork is `partition` also a complex expression in your case? 

```
where partition_date = DATE_PARSE('2021-07-01', '%Y-%m-%d') OR partition_date = DATE_PARSE('2021-07-012, '%Y-%m-%d')...
```
this is just one complex expression. What are other projections?

Reply:
Yes, it's also a complex expression, and the fact that the single complex expression in the inner query is referred to more than once in the outer filter caused it ineligible to be pushed down.
```
SELECT * FROM ( 
  SELECT date_parse(dt, '%Y-%m-%d') AS partition_date FROM t 
)
where partition_date = DATE_PARSE('2021-07-01', '%Y-%m-%d') OR partition_date = DATE_PARSE('2021-07-02, '%Y-%m-%d')...
```

Reply:
> There are other cases like where partition LIKE 'ab%' OR partition LIKE 'cd%'... or inequality conditions like where partition BETWEEN DATE '2000-01-01' AND '2000-12-31' OR partition BETWEEN '2020-01-01' AND '2020-12-31'... So I think it makes sense to relax this rule for all the common cases

Do you actually have this cases. If not, I would suggest implementing https://github.com/trinodb/trino/pull/2698 first. It strict improvement compared to heuristics here that could cause regressions

Reply:
Yes I've seen several use cases caused by this issue in our environment. And as I mentioned in the description of this PR, other people have reported this issue long time ago. So this is not uncommon. As we discussed earlier, we can add another limitation on "count" to avoid the possible expression explosion. I can make the update if you're ok with it.

Reply:
> As we discussed earlier, we can add another limitation on "count" to avoid the possible expression explosion

Yes, count limitation would be better than string expression length (this should be removed).